flash/image/classification/adapters.py

```diff
@@ -0,0 +1,513 @@
+# limitations under the License.
+import inspect
+import os
+from collections import defaultdict
+from functools import partial
+from typing import Any, Callable, List, Optional, Type
+
+import torch
+from pytorch_lightning import LightningModule
+from pytorch_lightning.plugins import DDPPlugin, DDPSpawnPlugin
+from pytorch_lightning.trainer.states import TrainerFn
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.warnings import WarningCache
+from torch.utils.data import DataLoader, IterableDataset, Sampler
+
+import flash
+from flash.core.adapter import Adapter, AdapterTask
+from flash.core.data.auto_dataset import BaseAutoDataset
+from flash.core.data.data_source import DefaultDataKeys
+from flash.core.model import Task
+from flash.core.registry import FlashRegistry
+from flash.core.utilities.imports import _LEARN2LEARN_AVAILABLE
+from flash.core.utilities.providers import _LEARN2LEARN
+from flash.core.utilities.url_error import catch_url_error
+
+warning_cache = WarningCache()
+
+
+if _LEARN2LEARN_AVAILABLE:
+    import learn2learn as l2l
+    from learn2learn.data.transforms import RemapLabels as Learn2LearnRemapLabels
+    from learn2learn.utils.lightning import Epochifier, TaskDataParallel
+else:
+
+    class Learn2LearnRemapLabels:
+        pass
+
+    class Epochifier:
+        pass
+
+    class TaskDataParallel:
+        pass
+
+
+class RemapLabels(Learn2LearnRemapLabels):
+    def remap(self, data, mapping):
+        # remap needs to be adapted to Flash API.
+        data[DefaultDataKeys.TARGET] = mapping(data[DefaultDataKeys.TARGET])
+        return data
+
+
+class NoModule:
+
+    """This class is used to prevent nn.Module infinite recursion."""
+
+    def __init__(self, task):
+        self.task = task
+
+    def __getattr__(self, key):
+        if key != "task":
+            return getattr(self.task, key)
+        return self.task
+
+    def __setattr__(self, key: str, value: Any) -> None:
+        if key == "task":
+            object.__setattr__(self, key, value)
+            return
+        setattr(self.task, key, value)
+
+
+class Model(torch.nn.Module):
+    def __init__(self, backbone: torch.nn.Module, head: Optional[torch.nn.Module]):
+        super().__init__()
+        self.backbone = backbone
+        self.head = head
+
+    def forward(self, x):
+        x = self.backbone(x)
+        if x.dim() == 4:
+            x = x.mean(-1).mean(-1)
+        if self.head is None:
+            return x
+        return self.head(x)
+
+
+class Learn2LearnAdapter(Adapter):
+
+    required_extras: str = "image"
+
+    def __init__(
+        self,
+        task: AdapterTask,
+        backbone: torch.nn.Module,
+        head: torch.nn.Module,
+        algorithm_cls: Type[LightningModule],
+        ways: int,
+        shots: int,
+        meta_batch_size: int,
+        queries: int = 1,
+        num_task: int = -1,
```

Comment:
@tchaton I think we might benefit from additional flexibility in the `num_task` argument. If you look at the [miniimagenet example here](https://github.com/learnables/learn2learn/blob/06893e847693a0227d5f35a6e065e6161bb08201/examples/vision/protonet_miniimagenet.py#L115-L140), you will see that they use different `num_task`'s when creating the `TaskDataset` for train, val, test

```python
train_tasks = l2l.data.TaskDataset(
    train_dataset, 
    task_transforms=train_transforms,  # note: it defaults to -1
)
test_tasks = l2l.data.TaskDataset(
    test_dataset,
    task_transforms=test_transforms,
    num_tasks=2000
)
valid_tasks = l2l.data.TaskDataset(
    valid_dataset,
    task_transforms=valid_transforms,
    num_tasks=200
)
```